pkg/kademlia/peer_discovery.go

```diff
@@ -117,3 +121,109 @@ func isDone(ctx context.Context) bool {
 		return false
 	}
 }
+
+// discoveryQueue is a limited priority queue for nodes with xor distance
+type discoveryQueue struct {
+	maxLen int
+	mu     sync.Mutex
+	added  map[storj.NodeID]int
+	items  []queueItem
+}
+
+// queueItem is node with a priority
+type queueItem struct {
+	node     *pb.Node
+	priority storj.NodeID
+	// TODO: switch to using pb.NodeAddress to avoid pointer to *pb.Node
+}
+
+// newDiscoveryQueue returns a items with priority based on XOR from targetBytes
+func newDiscoveryQueue(size int) *discoveryQueue {
+	return &discoveryQueue{
+		added:  make(map[storj.NodeID]int),
+		maxLen: size,
+	}
+}
+
+// Insert adds nodes into the queue.
+func (queue *discoveryQueue) Insert(target storj.NodeID, nodes ...*pb.Node) {
+	queue.mu.Lock()
+	defer queue.mu.Unlock()
+
+	unique := nodes[:0]
+	for _, node := range nodes {
+		nodeID := node.Id
```

Comment:
I think this logic should go after we have determined if we have hit our queue max length.

We don't want to say we have spoken to a node if we chop it off before we talk to it.

Reply:
fixed